File: src/main/java/com/letgo/book/infrastructure/persistence/mapping/HibernateBookMapper.java
package com.letgo.book.infrastructure.persistence.mapping;

import com.letgo.book.domain.Book;

final public class HibernateBookMapper {
    public static Book toDomainEntity(HibernateBookEntity entity) {
        Book book = Book.create(entity.id(), entity.title());
        book.retrieveEvents();
        return book;
    }

    public static HibernateBookEntity toOrmEntity(Book entity) {
        return new HibernateBookEntity(entity.id(), entity.title());
    }
}

File: src/main/java/com/letgo/book/infrastructure/persistence/mapping/HibernateBookMapper.kt
package com.letgo.book.infrastructure.persistence.mapping

import com.letgo.book.domain.Book
import com.letgo.book.domain.BookStatus
import com.letgo.book.domain.BookTitle

object HibernateBookMapper {
    fun toDomainEntity(entity: HibernateBookEntity): Book {
        val book: Book = Book.create(
            entity.id()!!, BookTitle.create(
                entity.title()!!.value()!!, entity.title()!!.createdAt()!!
            )
        )
        if (entity.status()!! == BookStatus.Edited) {
            book.changeTitle(book.title())
        }
        book.retrieveEvents()
        return book
    }

    fun toOrmEntity(entity: Book): HibernateBookEntity {
        return HibernateBookEntity(
            entity.id(),
            BookTitleEmbeddable(entity.title()),
            if (entity.hasBeenEdited()) BookStatus.Edited else BookStatus.Created
        )
    }
}
